pkg/kv/local/store.go

```diff
@@ -0,0 +1,229 @@
+		log.WithError(kv.ErrMissingKey).Warn("got empty key")
+		return nil, kv.ErrMissingKey
+	}
+
+	var value []byte
+	err := s.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get(composeKey(partitionKey, key))
+		if errors.Is(err, badger.ErrKeyNotFound) {
+			return kv.ErrNotFound
+		}
+		if err != nil {
+			log.WithError(err).Error("error getting key")
+			return err
+		}
+		value, err = item.ValueCopy(nil)
+		if err != nil {
+			log.WithError(err).Error("error getting value for key")
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		log.WithField("took", time.Since(start)).Trace("operation failed")
+		return nil, err
+	}
+	log.WithField("took", time.Since(start)).WithField("size", len(value)).
+		Trace("operation complete")
+	return &kv.ValueWithPredicate{
+		Value:     value,
+		Predicate: kv.Predicate(value),
+	}, nil
+}
+
+func (s *Store) Set(ctx context.Context, partitionKey, key, value []byte) error {
+	k := composeKey(partitionKey, key)
+	start := time.Now()
+	log := s.logger.WithField("key", string(k)).WithField("op", "set").WithContext(ctx)
+	log.Trace("performing operation")
+	if len(partitionKey) == 0 {
+		log.WithError(kv.ErrMissingPartitionKey).Warn("got empty partition key")
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		log.WithError(kv.ErrMissingKey).Warn("got empty key")
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		log.WithError(kv.ErrMissingValue).Warn("got nil value")
+		return kv.ErrMissingValue
+	}
+	err := s.db.Update(func(txn *badger.Txn) error {
+		return txn.Set(k, value)
+	})
+	if err != nil {
+		log.WithError(err).Error("error setting value")
+		return err
+	}
+	log.
+		WithField("took", time.Since(start)).
+		Trace("done setting value")
+	return nil
+}
+
+func (s *Store) SetIf(ctx context.Context, partitionKey, key, value []byte, valuePredicate kv.Predicate) error {
+	k := composeKey(partitionKey, key)
+	start := time.Now()
+	log := s.logger.WithField("key", string(k)).WithField("op", "set_if").WithContext(ctx)
+	log.Trace("performing operation")
+	if len(partitionKey) == 0 {
+		log.WithError(kv.ErrMissingPartitionKey).Warn("got empty partition key")
+		return kv.ErrMissingPartitionKey
+	}
+	if len(key) == 0 {
+		log.WithError(kv.ErrMissingKey).Warn("got empty key")
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		log.WithError(kv.ErrMissingValue).Warn("got nil value")
+		return kv.ErrMissingValue
+	}
+
+	err := s.db.Update(func(txn *badger.Txn) error {
+		item, err := txn.Get(k)
+		if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
+			log.WithError(err).Error("could not get key for predicate")
+			return err
+		}
+
+		if valuePredicate != nil {
+			if errors.Is(err, badger.ErrKeyNotFound) {
+				log.WithField("predicate", nil).Trace("predicate condition failed")
+				return kv.ErrPredicateFailed
+			}
+			val, err := item.ValueCopy(nil)
+			if err != nil {
+				log.WithError(err).Error("could not get byte value for predicate")
+				return err
+			}
```

Comment:
No need to copy value just to test the predicate.

Reply:
It's easier and more readable than using the `Value(fn func(val []byte) error)` error function